### PR TITLE
[DONTMERGE] Copy over everything from web-components

### PR DESCRIPTION
**Pull request by SvenDowideit**
so this is just a copy of the web-components code. I need to bring back:

(The _only_ difference in the first commit is the BrowserRouter to HashRouter change in index.tsx)

* [ ] Validator page
* [ ] Anchor page
* [ ] remove the IPC code for the solana web3 requests as they're in the renderer code now
* [ ] rearrange the dir structure so we can build both the electron app, and the web app from the one repo, and share the common-renderer components

Some of these may be a follow on PR...

### CODE REVIEW

**Comment by SvenDowideit** [comment]
@nathanleclaire ok, this is very ready for review&merge

**Comment by SvenDowideit** [comment]
mmm, yeah, i'm going to need you to tell me about the config serialisation - I'm confused why this compiles, considering i've removed the WBAccount type definition :/

**Comment by SvenDowideit** [comment]
and I've started to prototype "Transfer SOL" and "Airdrop SOL" buttons with pop-overs:

image

File: src/renderer/nav/ValidatorNetworkInfo.tsx [hunk 1]
```diff
@@ -1,45 +1,86 @@
...
-      }
-    };
-    window.electron.ipcRenderer.on('main', listener);
-    return () => {
-      window.electron.ipcRenderer.removeListener('main', listener);
-    };
-  }, []);
+    // TODO: set a spinner while waiting for response
+    fetchValidatorNetworkInfo(url)
+      .then((d) => setData(d))
```
Comment (nathanleclaire) [inline src/renderer/nav/ValidatorNetworkInfo.tsx:80]: `async` / `await`

File: src/renderer/store.ts [hunk 2]
```diff
@@ -1,6 +1,32 @@
...
+import throttle from 'lodash/throttle';
 
-export default configureStore({
-  reducer: mainReducer,
+// https://redux.js.org/usage/usage-with-typescript#define-slice-state-and-action-types
+// eslint-disable-next-line import/no-cycle
+import ValidatorReducer from './data/ValidatorNetwork/validatornetworkSlice';
+// https://redux.js.org/usage/usage-with-typescript#define-slice-state-and-action-types
+// eslint-disable-next-line import/no-cycle
+import SelectedAccountsListReducer from './data/SelectedAccountsList/selectedaccountsSlice';
```
Comment (nathanleclaire) [inline src/renderer/store.ts:10]: `selectedAccountsSlice`

File: src/renderer/components/ProgramChangeView.tsx [hunk 3]
```diff
@@ -1,92 +1,106 @@
...
-    paused: false,
+  const [changes, setChangesState] = useState<AccountInfo[]>([]);
+  const [paused, setPausedState] = useState<boolean>(false);
+
+  const displayList: string[] = []; // list of Keys
+  const pinnedAccount: PinnedAccountMap = {};
+
+  // this should be used to build the new list for display, not to mutate the changes list...
+  // eslint-disable-next-line array-callback-return
+  pinnedAccounts.map((key: string): void => {
```
Comment (nathanleclaire) [inline src/renderer/components/ProgramChangeView.tsx:80]: this should really be like a `forEach` or something right? `map` almost always is expected to return something

File: src/renderer/components/ProgramChangeView.tsx [hunk 4]
```diff
@@ -97,27 +111,22 @@ function ProgramChangeView(props: {
...
-    }
-  };
-  const unpause = () => {
-    if (pausedTimeoutRef.current !== 0) {
-      window.clearTimeout(pausedTimeoutRef.current);
-      pausedTimeoutRef.current = 0;
-    }
-    setPaused(false);
-  };
+  // // TODO: if {net} _maybe_ we want to unpause?
```
Comment (nathanleclaire) [inline src/renderer/components/ProgramChangeView.tsx:114]: rm

File: src/renderer/data/SelectedAccountsList/selectedaccountsSlice.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+// Define the initial state using that type
+let initialState: SelectedAccountsList = {
+  pinnedAccounts: [],
+  selectedAccount: '',
+  hoveredAccount: '',
+  editedAccount: '',
+  rootKey: '',
+};
+const loaded = loadState('selectedaccounts');
```
Comment (nathanleclaire) [inline src/renderer/data/SelectedAccountsList/selectedaccountsSlice.ts:25]: `selectedAccounts`, etc, couple other instances of that in this file if you could please address

File: src/renderer/data/accounts/getAccount.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+import { AccountInfo } from './accountsSlice';
+
+import { Net, netToURL } from '../ValidatorNetwork/validatornetworkSlice';
+
+export const BASE58_PUBKEY_REGEX = /^[1-9A-HJ-NP-Za-km-z]{32,44}$/;
+
+const HEXDUMP_BYTES = 512;
+
+// TODO: make a gloabal state of "current slot number"
```
Comment (nathanleclaire) [inline src/renderer/data/accounts/getAccount.ts:14]: i need this for transaction following (which I think I want to make replace program changes) as well

File: src/renderer/data/localstorage.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  }
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const saveState = (stateName: string, state: any) => {
+  try {
+    const serializedState = JSON.stringify(state);
+    localStorage.setItem(stateName, serializedState);
+  } catch (err) {
+    throw new Error("Can't save changes in local storage");
```
Comment (nathanleclaire) [inline src/renderer/data/localstorage.ts:19]: maybe add the inner `err` in?

File: src/renderer/nav/Account.tsx [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+import AccountView from '../components/AccountView';
+import ProgramChangeView from '../components/ProgramChangeView';
+import LogView from '../components/LogView';
+import { useAppSelector, useAppDispatch } from '../hooks';
+import { selectValidatorNetworkState } from '../data/ValidatorNetwork/validatornetworkSlice';
+import {
+  accountsActions,
+  selectAccountsListState,
+} from '../data/SelectedAccountsList/selectedaccountsSlice';
```
Comment (nathanleclaire) [inline src/renderer/nav/Account.tsx:17]: `selectedAccountsSlice`

File: src/renderer/nav/Account.tsx [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+import { getAccount } from '../data/accounts/getAccount';
+
+function Account() {
+  const dispatch = useAppDispatch();
+
+  const validator = useAppSelector(selectValidatorNetworkState);
```
Comment (nathanleclaire) [inline src/renderer/nav/Account.tsx:23]: i kinda like to group all the `useFoo`s together, but eh

**Review by nathanleclaire — CHANGES_REQUESTED** [review]

**Comment by nathanleclaire** [comment]
> mmm, yeah, i'm going to need you to tell me about the config serialisation - I'm confused why this compiles, considering i've removed the WBAccount type definition :/

You mean the config file persisted to disk?

**Comment by nathanleclaire** [comment]
So isn't using localStorage for accounts going to blow away all the users' stateful stuff when they update the Electron app? We'll fix that in our follow-up(s) to make the tool backend agnostic I guess?

**Comment by nathanleclaire** [comment]
... hm, mine complains about it 
Screen Shot 2022-04-04 at 2 03 05 PM

**Comment by nathanleclaire** [comment]
You're using Microsoft ESLint plugin right?

File: src/main/transactionLogs.ts [hunk 10]
```diff
@@ -6,6 +6,7 @@ import { netToURL } from '../common/strings';
 const logSubscriptions: LogSubscriptionMap = {};
 const subscribeTransactionLogs = async (
   event: Electron.IpcMainEvent,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/main/transactionLogs.ts:9]: it is a warning - and its an ipc based any, so its not worth changing the code

but I have fixed a stack of places where it was useful. so I prefer to put markers in the code that say "this is ok", and use warnings to tell me when to think more deeply about something

ie - I prefer to get to zero warnings

File: src/renderer/components/AccountListItem.tsx [hunk 11]
```diff
@@ -169,4 +218,8 @@ function AccountListItem(props: {
   );
 }
 
-export default AccountListItem;
+// AccountListItem.defaultProps = {
+//   queriedAccount: undefined,
+// };
+
+// export default AccountListItem;
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountListItem.tsx:225]: yeah - I was operating under the assumption that typescript would error out when things didn't match up - and now i'm 100% weirded out too

i need an npm command that will tell me everything, and i currently don't have one.

File: src/renderer/App.tsx [hunk 12]
```diff
@@ -186,191 +125,79 @@ function Topbar() {
...
-                      />
-                      <TooltipNavItem
-                          to="/validatornetworkinfo"
-                          title="Network Info"
-                          tooltipMessage="Network Info"
-                          eventKey="validatornetworkinfo"
-                          icon={faNetworkWired}
-                          iconsize="xl"
-                      /> */}
+              to="/"
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/App.tsx:128]: made a function :p

File: src/renderer/components/AccountNameEditable.tsx [hunk 13]
```diff
@@ -26,13 +30,14 @@ function AccountNameEditable(props: {
...
       editingStopped={() => dispatch(accountsActions.setEdited(''))}
       handleOutsideClick={() => {
-        analytics('updateAccountName', {});
-        window.electron.ipcRenderer.updateAccountName({
-          net,
-          pubKey,
-          humanName: ref.current.value,
-        });
+        analytics('updateAccountName', { net, pubKey });
+        // updateAccountName({
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountNameEditable.tsx:34]: ah, actually, no, I have no plan on this - I just havn't figured it out yet (which is why there are zero implementations of editing and storing the human name yet)

* [ ] SVEN - TODO

File: src/renderer/components/AccountNameEditable.tsx [hunk 14]
```diff
@@ -26,13 +30,14 @@ function AccountNameEditable(props: {
...
+        //     net,
+        //     pubKey,
+        //     humanName: ref.current.value,
+        // })
       }}
+      // eslint-disable-next-line react/jsx-props-no-spreading
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountNameEditable.tsx:40]: warnings are _not-OK_, they're a warning that you may have something sub-optimal, and you should either change it, or mark it as "meh, not worth fixing"

in this case, I am not using this component yet, so punted.

File: src/renderer/components/AccountView.tsx [hunk 15]
```diff
@@ -1,24 +1,97 @@
...
+  const { pubKey } = props;
+  const { net } = useAppSelector(selectValidatorNetworkState);
+
+  const [account, setSelectedAccountInfo] = useState<AccountInfo | undefined>(
+    undefined
+  );
+
+  const updateAccount = useCallback(() => {
+    if (pubKey) {
+      getAccount(net, pubKey)
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountView.tsx:51]: there was a reason, i'll see if i can recreate it - tho its looking like i dramatically changed getAccounts since.

File: src/renderer/nav/Account.tsx [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    string | undefined
+  >(selectedAccount);
+
+  const attemptAccountAdd = (pubKey: string) => {
+    dispatch(accountsActions.setSelected(pubKey));
+    getAccount(net, pubKey)
+      .then((res) => {
+        // eslint-disable-next-line promise/always-return
+        if (res) {
+          if (res instanceof Error) {
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/nav/Account.tsx:38]: lol, try catch? what's that? I'm a go programmer dammit!

File: src/renderer/nav/Account.tsx [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+          setSelectedAccountInfo(pubKey);
+        }
+      })
+      /* eslint-disable no-console */
+      .catch(console.log);
+  };
+
+  // TODO: the borders should eventually be resizable
+  return (
+    <Stack className="almostvh100">
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/nav/Account.tsx:50]: make a lint error for that...

File: src/renderer/components/AccountView.tsx [hunk 18]
```diff
@@ -1,24 +1,97 @@
...
+        })
+        /* eslint-disable no-console */
+        .catch(console.log);
+    } else {
+      setSelectedAccountInfo(undefined);
+    }
+  }, [net, pubKey]);
+
+  useEffect(() => {
+    updateAccount();
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountView.tsx:71]: na, it was a wrapper because i was reusing it in more than one place

[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountView.tsx:71]: the useEffect is for triggering when the user changes the network / selected pubKey - even if they disable the interval (which i suspect will come)

File: src/renderer/components/AccountView.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountView.tsx:51]: I think https://www.benmvp.com/blog/successfully-using-async-functions-useeffect-react/ shows why not, and IMO, `.then()` is a heck of a lot better than the nested `async`

File: src/renderer/components/AccountListItem.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountListItem.tsx:225]: OH - I'm not using AccounListItem anymore - its replaced entirely, will delete

File: src/renderer/components/AccountView.tsx [hunk 19]
```diff
@@ -91,10 +164,10 @@ function AccountView(props: { account: WBAccount }) {
               </table>
             </div>
             <div className="col-auto">
-              <RandomArt
+              {/* <RandomArt
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountView.tsx:149]: BUTILIKEIT!

File: src/renderer/components/LogView.tsx [hunk 20]
```diff
@@ -21,24 +48,31 @@ function LogView() {
...
     return () => {
-      window.electron.ipcRenderer.removeAllListeners('transaction-logs');
-      window.electron.ipcRenderer.unsubscribeTransactionLogs({
-        net,
-      });
+      const sub = logSubscriptions[net];
+      sub.solConn
+        .removeOnLogsListener(sub.subscriptionID)
+        // eslint-disable-next-line promise/always-return
+        .then(() => {
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/LogView.tsx:61]: https://www.benmvp.com/blog/successfully-using-async-functions-useeffect-react/

File: src/renderer/components/ProgramChange.tsx [hunk 21]
```diff
@@ -1,76 +1,105 @@
...
-  const imported = pubKey in importedAccounts;
-  const [importing, setImporting] = useState(false);
+  const { pubKey, net, attemptAccountAdd, pinned, pinAccount } = props;
+
+  const [change, setChangeInfo] = useState<AccountInfo | undefined>(undefined);
+
+  const updateAccount = useCallback(() => {
+    if (pubKey) {
+      getAccount(net, pubKey)
+        .then((res) => {
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/ProgramChange.tsx:29]: https://www.benmvp.com/blog/successfully-using-async-functions-useeffect-react/

File: src/renderer/components/ProgramChange.tsx [hunk 22]
```diff
@@ -1,76 +1,105 @@
...
   };
   return (
-    <>
+    <Container
+      onClick={() => {
+        attemptAccountAdd(pubKey, false);
+      }}
+    >
+      <td
+        onClick={() => {
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/ProgramChange.tsx:72]: yeah, its common and i hate it.

File: src/renderer/data/accounts/getAccount.ts [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+export const updateCache = (account: AccountInfo) => {
+  cache.set(`${account.net}_${account.pubKey}`, account);
+};
+
+export async function getAccount(
+  net: Net,
+  pubKey: string
+): Promise<AccountInfo | Error | undefined> {
+  // logger.info("getAccount", {pubKey});
+  const cachedResponse = cache.peek(`${net}_${pubKey}`);
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/data/accounts/getAccount.ts:39]: the purpose to make trivial to keep any view updated - irrespective of where - and to allow the programmer not to worry about there being load.

basically - its a comms abstraction - watch the validator staked account some time - its SOL view gets updated in both the program changes and account view - but with this cache, it doesn't do the network request for getAccount, because the info is updated from program changes often enough.

basically, there are at least 3 different ways of getting the info from the validator, and at least 2 (and soon more) ways we may display it live - and this way, the views of that info uses as few network requests as possible to show an updated and consistent value.

I'm also planning on having a multi-browser-tab cache, so we don't add extra network and parsing load when the user does that.

File: src/renderer/data/accounts/programChanges.ts [hunk 24]
```diff
@@ -0,0 +1,122 @@
+import * as sol from '@solana/web3.js';
+import { Net, netToURL } from '../ValidatorNetwork/validatornetworkSlice';
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/data/accounts/programChanges.ts:2]: no, `types` is a very bad global meme, and needs to go away. having a big global types file increases coupling risks as codebases grow.

File: src/renderer/index.css [hunk 25]
```diff
@@ -0,0 +1,13 @@
+body {
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/index.css:1]: maybe - but App.scss is only for electron specific things - we're about to have dual deployment.

File: src/renderer/components/ProgramChangeView.tsx [hunk 26]
```diff
@@ -127,139 +136,134 @@ function ProgramChangeView(props: {
...
+                    />
+                  </div>
+                </DropdownButton>
+              </OutsideClickHandler>
+            </Dropdown>
+            <Button size="sm" onClick={() => setPausedState(!paused)}>
+              {paused ? 'unpause' : 'pause'}
+            </Button>
+            <Button size="sm" disabled>
+              Import account
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/ProgramChangeView.tsx:201]: ui placeholder - trying to see where it makes sense in the UI

File: src/renderer/components/AirDropSolButton.tsx [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+  if (!pubKeyVal) {
+    pubKeyVal = 'paste';
+  }
+
+  const [sol, setSol] = useState<string>('0.01');
+  const [toKey, setToKey] = useState<string>(pubKeyVal);
+
+  return (
+    <Popover id="popover-basic">
+      <Popover.Header as="h3">Airdrop SOL</Popover.Header>
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AirDropSolButton.tsx:33]: I'm exploring the UI concepts - give it time.

File: src/renderer/data/ValidatorNetwork/validatornetworkSlice.ts [hunk 28]
```diff
@@ -0,0 +1,69 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
+
+// https://redux.js.org/usage/usage-with-typescript#define-slice-state-and-action-types
+// eslint-disable-next-line import/no-cycle
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/data/ValidatorNetwork/validatornetworkSlice.ts:4]: read the URL.

File: src/renderer/data/accounts/account.ts [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+  // await connection.confirmTransaction(airdropSignature);
+}
+
+export async function transferSol(
+  fromKey: string,
+  toKey: string,
+  solAmount: string
+) {
+  console.log(`move ${solAmount} from ${fromKey} to ${toKey}`);
+  await new Promise((resolve) => setTimeout(resolve, 2000));
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/data/accounts/account.ts:26]: that was what i thought too, and then i got it to crash :)

File: src/renderer/data/accounts/account.ts [hunk 30]
```diff
@@ -0,0 +1,95 @@
+import * as web3 from '@solana/web3.js';
+// import { struct, u32, ns64 } from '@solana/buffer-layout';
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/data/accounts/account.ts:2]: oh, not yet, I'm contemplating the two different ways to write the transfer code - the commented out version is what's in the solana docs, and i want to work out why.

File: src/renderer/data/accounts/account.ts [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+  const connection = new web3.Connection(netToURL(net));
+
+  const airdropSignature = await connection.requestAirdrop(
+    payer.publicKey,
+    web3.LAMPORTS_PER_SOL
+  );
+
+  await connection.confirmTransaction(airdropSignature);
+
+  // const allocateTransaction = new web3.Transaction({
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/data/accounts/account.ts:43]: seeabove.

File: src/renderer/data/localstorage.ts [hunk 32]
```diff
@@ -0,0 +1,21 @@
+export const loadState = (stateName: string) => {
```
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/data/localstorage.ts:1]: I'd like consistency of only one of snake, camel, skewer, whatever - not a mix of different ones in different contexts.

and either way, if it is important, make a linter, and a prettify rule - ala gofmt.

File: src/renderer/components/AccountNameEditable.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountNameEditable.tsx:34]: mmm, i think i'd like to deal with this in a later PR

File: src/renderer/data/SelectedAccountsList/selectedaccountsSlice.ts (hunk 5 above)
Comment (SvenDowideit) [reply src/renderer/data/SelectedAccountsList/selectedaccountsSlice.ts]: slice names seem to be done as all lowercase - so it seems more consistent to me to do the same thing for saving the slice into the localState

File: src/renderer/nav/Account.tsx (hunk 9 above)
Comment (SvenDowideit) [reply src/renderer/nav/Account.tsx:22]: yeah, i'm still struggling with what to group and what not, its way too arbitrary.

**Comment by SvenDowideit** [comment]
I need my cli build tooling to at minimum have the errors and warnings the editor shows - in this PR's case, I basically missed the main issues because i didn't open them - I was expecting (really stupidly) that the `npm` would tell me about issues in files i didn't have open - and was using `npx eslint --fix src` to tell me where I needed to fix next.

File: src/renderer/data/accounts/account.ts (hunk 29 above)
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/data/accounts/account.ts:26]: eh, whatever, this is a placeholder for something that needs to be completed in a later PR - need to safely store account private keys first :(

File: src/renderer/components/AccountNameEditable.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/AccountNameEditable.tsx:34]: oh, and no, there aren't two implementations once this PR merges.

File: src/renderer/components/ProgramChangeView.tsx (hunk 3 above)
Comment (SvenDowideit) [reply src/renderer/components/ProgramChangeView.tsx:77]: yeah, this one has been confusing me - dunno why brb

File: src/renderer/data/SelectedAccountsList/selectedaccountsSlice.ts (hunk 5 above)
Comment (SvenDowideit) [reply src/renderer/data/SelectedAccountsList/selectedaccountsSlice.ts]: mmm, nope, can't find them, we'll need to talk about this so i have a chance of groking.

File: src/renderer/components/ProgramChangeView.tsx (hunk 26 above)
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/components/ProgramChangeView.tsx]: pushed it up out of the program changes

File: src/renderer/index.css (hunk 25 above)
[the remark this answers is not in the document]
Comment (SvenDowideit) [reply src/renderer/index.css:1]: meh, this one i don't have a stake in, and as i created it, i'm killing it.

**Comment by nathanleclaire** [comment]
Alright I suspect you will get these CI errors ironed out soon, if you get the green today ahead and merge, if not, give it a second and I'll look at them in AM to see if I can help troubleshoot! LGTM once `npm exec tsc` fixed

**Review by nathanleclaire — APPROVED** [review]

**Comment by SvenDowideit** [comment]
half way through re-making the config slice on the render side - its close, but not complete - tomorrow!

**Comment by SvenDowideit** [comment]
ok, what remains needs more careful work - they're all because I moved all the account selection state info to client side state only using localStorage, and we need to work out a good time, place and method for pushing that back to the main loop

```
sven@p1:~/src/cryptoworkbench/solana-workbench$ npm exec tsc
src/main/accounts.ts:7:3 - error TS2305: Module '"../types/types"' has no exported member 'AccountsResponse'.

7   AccountsResponse,
    ~~~~~~~~~~~~~~~~

src/main/accounts.ts:9:3 - error TS2305: Module '"../types/types"' has no exported member 'GetAccountResponse'.

9   GetAccountResponse,
    ~~~~~~~~~~~~~~~~~~

src/main/accounts.ts:14:3 - error TS2305: Module '"../types/types"' has no exported member 'WBAccount'.

14   WBAccount,
     ~~~~~~~~~

src/main/db.ts:5:26 - error TS2305: Module '"types/types"' has no exported member 'WBAccount'.

5 import { Net, ConfigMap, WBAccount } from 'types/types';
                           ~~~~~~~~~


Found 4 errors in 2 files.

Errors  Files
     3  src/main/accounts.ts:7
     1  src/main/db.ts:5
```

I kinda wonder if the json file creation maybe should happen client side too - so the web version can allow the user to download the json file to give solana - but its 10pm, so i'll bunt to tomorrow :D

**Comment by SvenDowideit** [comment]
REBASING - squashing first, cos rebasing 52 times is horrible

**Comment by SvenDowideit** [comment]
wow, so 

```
sven@p1:~/src/cryptoworkbench/solana-workbench$ npm run package

> package
> ts-node ./.erb/scripts/clean.js dist && npm run build && electron-builder build --publish never


> build
> concurrently "npm run build:main" "npm run build:renderer"

[0] 
[0] > build:main
[0] > cross-env NODE_ENV=production webpack --config ./.erb/configs/webpack.config.main.prod.ts
[0] 
[1] 
[1] > build:renderer
[1] > cross-env NODE_ENV=production webpack --config ./.erb/configs/webpack.config.renderer.prod.ts
[1] 
[0] [webpack-cli] Failed to load '/home/sven/src/cryptoworkbench/solana-workbench/.erb/configs/webpack.config.main.prod.ts' config
[0] [webpack-cli] .erb/configs/webpack.config.main.prod.ts(9,38): error TS7016: Could not find a declaration file for module 'webpack-bundle-analyzer'. '/home/sven/src/cryptoworkbench/solana-workbench/node_modules/webpack-bundle-analyzer/lib/index.js' implicitly has an 'any' type.
[0]   Try `npm i --save-dev @types/webpack-bundle-analyzer` if it exists or add a new declaration (.d.ts) file containing `declare module 'webpack-bundle-analyzer';`
[0] 
[0] npm run build:main exited with code 2
[1] [webpack-cli] Failed to load '/home/sven/src/cryptoworkbench/solana-workbench/.erb/configs/webpack.config.renderer.prod.ts' config
[1] [webpack-cli] .erb/configs/webpack.config.renderer.prod.ts(9,38): error TS7016: Could not find a declaration file for module 'webpack-bundle-analyzer'. '/home/sven/src/cryptoworkbench/solana-workbench/node_modules/webpack-bundle-analyzer/lib/index.js' implicitly has an 'any' type.
[1]   Try `npm i --save-dev @types/webpack-bundle-analyzer` if it exists or add a new declaration (.d.ts) file containing `declare module 'webpack-bundle-analyzer';`
[1] 
[1] npm run build:renderer exited with code 2
```

so another command that can trip you up, even when `npm start` is happy :(

**Comment by nathanleclaire** [comment]
I think you need to do `npm install`? Probably cause of the updated changes in #75 . This works fine on my computer w/ your branch and `npm run package` is part of the CI.

Not much we can do about errors on run/package if the correct modules haven't been installed.

**Comment by nathanleclaire** [comment]
LGTM
